### ISSUE TITLE
Could not determine loop bound at compile-time

### ISSUE THREAD

**Issue by huangliang-hl** [noir-lang/noir#2653]
### Aim

Implementing the slice method for the Tensor struct through recursion.

### Expected Behavior

This slice should be able to compile and run successfully, returning an array of a specified length.

### Bug

```
error: Could not determine loop bound at compile-time
   ┌─ /code/convolution/test/tensor_slice/src/main.nr:96:25
   │
96 │             for n in 0..sub_result_len {
   │                         --------------
   │
   = Call stack:
     1. /code/convolution/test/tensor_slice/src/main.nr.nr:109:22
     2. /code/convolution/test/tensor_slice/src/main.nr.nr:58:32
     3. /code/convolution/test/tensor_slice/src/main.nr.nr:95:30
     4. /code/convolution/test/tensor_slice/src/main.nr.nr:96:25

Error: Aborting due to 1 previous error

Location:
    crates/nargo_cli/src/cli/mod.rs:91:5
```

### To Reproduce

1.git clone https://github.com/storswiftlabs/convolution.git
2.cd convolution/test/tensor_slice
3.nargo execute

### Installation Method

Compiled from source

### Nargo Version

nargo 0.11.1 (git version hash: 9da822f59923a9953894c43afd1ddbeffa871dbf, is dirty: false)

### Would you like to submit a PR for this Issue?

No

**Comment by huangliang-hl** [comment]
Running with version 0.9.0 of nargo results in another error:
```
The application panicked (crashed).
Message:  assertion failed: `(left == right)`
  left: `Id(51)`,
 right: `Id(84)`: Expected two blocks to join to the same block
Location: crates/noirc_evaluator/src/ssa/opt/flatten_cfg/branch_analysis.rs:78

This is a bug. We may have already fixed this in newer versions of Nargo so try searching for similar issues at https://github.com/noir-lang/noir/issues/.
If there isn't an open issue for this bug, consider opening one at https://github.com/noir-lang/noir/issues/new?labels=bug&template=bug_report.yml
```

**Comment by jfecher** [comment]
Apologies for the delay in responding to this issue. As of the most recent commit on master, the original error (but not the follow up comment) is still present. The compiler currently has trouble dealing with mutable values used in for loop ranges since being used in a for loop range means the value must be known at compile-time - specifically before the loop unrolling step - while being mutable means we require a separate pass to determine it's current value at each point in the program.

The program you gave is a somewhat tough case in this where we'd have to unroll the first loop where `sub_result_len` is assigned to, then run the pass to handle mutability, then resume the unrolling pass now that the value of `sub_result_len` is known. This isn't possible currently, hence the error.

**Comment by TomAFrench** [comment]
This seems like it would have been fixed by #4779. Feel free to reopen if you're still running into issues.
